pyscraper/regmem/commons/process.py

```diff
@@ -0,0 +1,383 @@
+        response.raise_for_status()
+        data = response.json()
+        target_items = data["totalResults"]
+        all_items.extend(data["items"])
+        bar.total = target_items
+        bar.update(take)
+        skip += take
+        if skip >= target_items:
+            continue_fetching = False
+
+    return all_items
+
+
+def get_list_of_registers(quiet: bool = False):
+    url = REGISTER_BASE + "api/v1/Registers"
+    items = recursive_fetch(url, quiet=quiet)
+    registers = TypeAdapter(list[PublishedRegister]).validate_python(items)
+    return registers
+
+
+def move_subitems_under_parent(
+    interests: list[PublishedInterest], wider_list: list[PublishedInterest]
+) -> list[PublishedInterest]:
+    all_item_ids = []
+
+    parent_items = {x.id: x for x in interests if x.parent_interest_id is None}
+    wider_lookup = {x.id: x for x in wider_list}
+    all_item_ids.extend(parent_items.keys())
+
+    for interest in interests:
+        if interest.parent_interest_id is not None:
+            if interest.parent_interest_id not in parent_items:
+                parent_interest = wider_lookup[interest.parent_interest_id]
+                parent_items[parent_interest.id] = parent_interest
+                all_item_ids.append(parent_interest.id)
+            parent = parent_items[interest.parent_interest_id]
+            parent.child_items.append(interest)
+            all_item_ids.append(interest.id)
+
+    if len(all_item_ids) < len(interests):
+        raise ValueError("Not all items are accounted when moving to subitems.")
+
+    return list(parent_items.values())
+
+
+def convert_list_to_regmem_hierarchy(
+    interests: list[PublishedInterest],
+) -> list[Member]:
+    """
+    Here we're converting a mixed list of interests for different people and categories
+    into a hierarchy of members > categories > interests > subinterests.
+
+    """
+
+    interests.sort(key=lambda x: x.member.id)
+
+    members: list[Member] = []
+
+    for _, member_interests in groupby(interests, key=lambda x: x.member.id):
+        list_of_member_interests = list(member_interests)
+        member = list_of_member_interests[0].member
+
+        list_of_member_interests.sort(key=lambda x: x.category.id)
+
+        categories = []
+
+        for _, category_interests in groupby(
+            list_of_member_interests, key=lambda x: x.category.id
+        ):
+            list_of_category_interests = list(category_interests)
+            category = list_of_category_interests[0].category
+
+            category.interests = move_subitems_under_parent(
+                list_of_category_interests, wider_list=list_of_member_interests
+            )
+
+            categories.append(category)
+
+        member.categories = categories
+
+        members.append(member)
+
+    members.sort(key=lambda x: x.id)
+
+    # given we pull in parent interests when not in the category, we need to check we haven't duplicated
+
+    existing_ids = []
```

Comment:
Should these be sets, if I'm understanding it right? Could get same ID more than once

Reply:
Yeah, it's not big enough for a real performance penalty but makes it clearer they should be unique.